Source/interfac.cpp

```
@@ -650,9 +673,10 @@ void ShowProgress(interface_mode uMsg)
 	while (true) {
 		CheckShouldSkipRendering();
 		SDL_Event event;
-		// We use the real `PollEvent` here instead of `FetchMessage`
-		// to process real events rather than the recorded ones in demo mode.
-		while (PollEvent(&event)) {
+		// We call `CustonEventHandler.poll` directly instead of
+		// calling the `FetchMessage` to process real events rather
+		// than the recorded ones in demo mode.
+		while (CurrentEventHandler.poll(&event)) {

```

Can you have a look at #7712? It appears to me that you've replaced every call to `PollEvent()` with `SDL_PollEvent()`, which I expect will reintroduce this bug for a second time.

---

Right, I accidentally replaced `PollEvent()` here and in `FetchMessage_Real()` with `SDL_PollEvent` or `SDL_PeepEvent()`. I missed that `PollEvent()` also handles controller states and motion by calling:

```
UnlockControllerState(*event);
ProcessControllerMotion(*event);
```

Good catch, thanks.


I updated the code by splitting the commit in question on two for simpler review:

- The first one splits the `EventHandler` into the actual handler and poller callback. This commit should not change the logic of event processing and is just preparation for the next commit. The `PollEvent()` calls from `input.h` should be kept as they are.
- The second commit introduces `SDL_PeepEvent()` instead of `SDL_PollEvent()` for all the message handlers which are called from the `ShowProgress()` routine. This is actually a fix for the problem described.

The main idea is that `ShowProgress()` handles only custom events and `SDL_QUIT`, leaving other events in the queue for future processing. `PollEvent()` is replaced with `PollEventCustom()`, which performs the same input processing but uses the `SDL_PeepEvent()` call for polling.

Please, take a look.

---

Why do we want to leave events in the queue rather than process and ignore most of them? Wouldn't it be odd if I pressed some keys in the loading screen and suddenly they got processed all at once once I get in-game?

---

Sharp :) There are nuances. When your hero steps on a warp (e.g. stairs to another level) a few other events can be already queued. 

Theoretically an agent can avoid movements to certain objects, which lead to a warp but the whole issue comes from the fact, that while training (at least with reinforcement learning) agent starts exploring knowing nothing about the environment. And that can be a part of the training policy to teach an agent to avoid warps thought the negative rewards (thus reinforcement learning). What I'm trying to say is that an agent *has* to step in a warp in order to "understand" the consequences. 

Another thing is that the game engine provides very little detail on the current state of the game itself. Is game loading? Saving? Warping a hero? This limits interaction. Sometimes an agent has to derive the state indirectly. Example: agent explores, queues events, steps on a warp, does not know that this is a warp and continues sending events, suddenly the whole environment changes and dungeon level counter increases. Seeing the difference in level counter an agent deducts about warping, but it is impossible to say how many events were lost.

I apologize for these long texts. I just want to demonstrate that I thoughtfully tried to solve this problem in the agent, without touching the code in the game engine itself. Eventually I realized that it would be better to make a patch and push it here.

---

Right, I forgot about low-powered cpus, that can be slow indeed. 

From the human interaction perspective, as you correctly noticed, movements are slow, and once you have queued events during a gameplay (not while the load screen) you probably don't expect them to be silently dropped. Example: before a character steps on the trigger tile you press save. That has happened not while the load screen, but even before. You expect the game to be saved. Or loaded. But that does not happen.

What if we disable all new events while we are in the `ShowProgress()` with the `SDL_EventState(..., SDL_IGNORE)`, but keeping processing the queue as I have proposed by this path? This drains the queue, keeping all new events being ignored.

---

Sorry just catching up to this rather long discussion. RE this:

> Another thing is that the game engine provides very little detail on the current state of the game itself. Is game loading? Saving? Warping a hero? 

Given that a human can see this information, I think it'd make sense to provide it from the engine to the agent?

---

> Okay, I thought this might be what you were talking about, but I suppose you confused me by talking about human players. Input events do not make their way into the queue instantaneously upon action by the human player. 

Oh, my apologies, I did not mean to confuse you by jumping from human interaction to a simulation. What I wanted to demonstrate is that if an SDL event comes at the right time (right after the custom event is pushed), it will be lost. The logic is the following:

1. Player stopped on a trigger tile.
2. On each game loop iteration, `CheckTriggers()` is called, and once the player is on the trigger tile, a new custom event is pushed.
3. The custom event is processed in the `RunGameLoop()`, while other events are discarded.

Once any SDL event lands in the queue between steps 2 and 3, it will be lost. This logic is demonstrated by my snippet.

> The process for humans is naturally very asynchronous between pixels changing on the display, eyes receiving photons from the screen
...
> The human player will never notice your discrepancy amidst all the asynchronous "guesswork" their brain has to do to make all of this happen. 

You're right that human input is asynchronous and delayed at many stages from perception to motor action to event dispatch. But regardless of a delay along the way, the final destination of the event is a queue, and what matters is the actual moment when an event added to the queue. There's a small but real window where the SDL event generated by a human can be silently lost if it happens to land right after the custom event is pushed to the queue.

So the problem does affect human players, just not consistently or predictably, which arguably makes it worse, since these bugs are the hardest to diagnose. That's why I believe it's worth addressing.

> That solution is indeed better for the human player because there's no way they'd be able to tell the difference. That said, I am curious why you think this solution would fix the problem for your bot.

My agent experienced the problem with lost events and hanging in infinite wait. This PR ensures that no events are lost, and extensive RL training (I usually run up to a hundred instsances simultaneously on a machine, which of course increases the chance to hit an issue) does not reproduce an issue for months.

> How is the bot sending signals to the process? If it is asynchronous like the human player, then you can't fully prevent race conditions and must implement a different type of feedback.

It is asynchronous, and the only guarantee needed is that no events are lost, which this PR addresses.

> This is counter-productive. We were talking about the save action in the sense that you didn't mention ESC until just now. My point doesn't change based on this new information. The fact that "technically it also applies to the ESC key" makes no difference because your logic still doesn't apply to all keystrokes, mouse events, gamepad button events, window resize events, etc.

I'm not sure I follow. I described the problem of lost SDL events - any event that reaches the queue in the right moment. In my last example snippet, I mentioned the 'ESC' key sequence because it clearly shows the menu that a user would expect to see after pressing it, and users should notice if the sequence is lost and the menu does not appear. That was the goal of the shippet - to demonstrate.

What's worse, all press inputs (applies to all sorts of input devices: mouse, keyboard, or gamepad) generate two events: DOWN and UP events. Let's take a mouse button press as an example. If a DOWN event lands the queue but the UP event is lost, the player warps to another level and continues moving.

Here is another code snippet that shows the sequence of split mouse events:

```
--- a/Source/player.cpp
+++ b/Source/player.cpp
@@ -2860,12 +2862,34 @@ StartNewLvl(Player &player, interface_mode fom, int lvl)
        if (&player == MyPlayer) {
                player._pmode = PM_NEWLVL;
                player._pInvincible = true;
+
                SDL_Event event;
+
+               // Let's imagine the MOUSE BUTTON DOWN event lands in the
+               // queue right before the custom event was pushed.
+               event.type = SDL_MOUSEBUTTONDOWN;
+               event.button.button = SDL_BUTTON_LEFT;
+               event.button.state = SDL_PRESSED;
+               SDL_PushEvent(&event);
+
                CustomEventToSdlEvent(event, fom);
                SDL_PushEvent(&event);
+
+               // and MOUSE BUTTON UP event lands in the queue right after
+               // the custom event was pushed. This event will be discarded
+               // by the `ShowProgress()` filter and the player never stops
+               // moving until next MOUSE BUTTON sequence.
+               event.type = SDL_MOUSEBUTTONUP;
+               event.button.button = SDL_BUTTON_LEFT;
+               event.button.state = SDL_RELEASED;
+               SDL_PushEvent(&event);
```

If you run this on the latest master, you should see something like the following https://youtu.be/90Q2Txev9c0. The player warps and continues moving unless another mouse button is pressed.

I see no technical or probabilistic reasons why a human could not, accidentally or intentionally, send a mouse click, which (by a lucky chance) lands in the queue when a player stops on a trigger tile exactly as the snippet above demonstrates. This vaguely reminds me of the famous item duplication bug, which also highly depends on the correct timing of events.

---

> Since a human can see this information, I think it'd make sense to provide it from the engine to the agent?

I assume this information can be extracted from the engine using some static variables, which indirectly help derive the game state. This needs to be carefully developed. What I was saying is that there is no such thing as a game state, which complicates the whole automation.

---

>  That is why the human player can't tell the difference between a key press that gets queued between 2 and 3 and a keypress that gets queued after 3. Whether we change the behavior of the logic that processes events from the queue or not, the human experience does not change.

Absolutely agree - there's no visual difference for a human to spot regarding when an event is executed. But if a human presses ESC (or any other key action to your choice) expecting the game to show a menu and this doesn't happen, then that's an issue. Of course, we could speculate that the key was pressed almost simultaneously with the moment when a loading screen should show up, but in my opinion, this isn't proper behavior for an event queue. Lost events are a nightmare to debug and fix.

> I don't know what bugs you're talking about.

Fair enough. It’s a weird edge case and not easy to catch. I guess it just comes down to different styles: some people are more reactive, others more proactive about fixing stuff.

> Processing events in the order in which they are queued is not a bug. That's how the event queue is supposed to work.

Correct, the order is not an issue, but discarding events is.

> If it is asynchronous, then there can be no guarantee. Even if you change the cutoff point from 2 to 3, the bot has no guarantee that the event it sent before it observed 3 actually made it to the queue before the moment 3 actually occurred. It sounds to me like you'd be better off implementing a timeout based on feedback from the game so the bot can recognize, the same way a human player would, that the action it intended to execute wasn't executed.

The agent does not concern itself with when an event is executed relative to the queue, this is out of the scope of this discussion. What the agent expects is that the event eventually reaches completion, regardless of event position in the queue. Unfortunately, with the current queue design, this is not guaranteed, thus this PR.

> I don't suppose you've tried changing the order of those events?

No, I have not.

> I ran some tests just now, and it doesn't seem to matter whether the SDL_RELEASED event is processed before or after the custom event. Try running this code.

Interesting. That's seems to me a regression, I was able to bisect to this one:

   3e6b501d82e7 ("Fix pathfinding and increase player path limit")

Cc: @glebm

> The bug you've found has nothing to do with whether the engine "discards" the mouse up event.

Ooops. But still, the bug I've found doesn't prove the opposite, that discard won't happen, right? :) And peeking into the queue to see whether the UP event exists or not won't be difficult.

---

> Of course, we could speculate that the key was pressed almost simultaneously with the moment when a loading screen should show up, but in my opinion, this isn't proper behavior for an event queue. Lost events are a nightmare to debug and fix.

This is not speculation. That's literally the thing you're describing. The event _has_ to be queued in the same "display frame" as the custom event is queued, otherwise they would be processed in separate iterations of the main loop. If they are processed in separate iterations of the main loop, then there is no choice; you have to process them in the order they appear in the queue. If they appear in separate iterations of the rendering loop, then you may as well process them in the order they appear in the queue, because there is no difference as far as the engine is concerned.

> Correct, the order is not an issue, but discarding events is.

> The agent does not concern itself with when an event is executed relative to the queue, this is out of the scope of this discussion. What the agent expects is that the event eventually reaches completion, regardless of event position in the queue. Unfortunately, with the current queue design, this is not guaranteed, thus this PR.

Perhaps I haven't been clear enough about this. No events are actually "lost" or "discarded" in the process of transitioning to the loading screen. The mouse up event that you claim was lost is actually being processed by the `ProgressEventHandler()` function, which invokes the `DisableInputEventHandler()` function, which in turn sets `sgbMouseDown = CLICK_NONE`. The only way in which it is "lost" is the fact that it doesn't get processed by the `GameEventHandler()` function. I am arguing that this is proper because that handler executes in an entirely different context where a mouse up event might open a UI menu or something. _It only stands to reason that a mouse up event would work differently on the loading screen, which doesn't have a UI menu to open._

---

Let me sum it all up so we can wrap this long discussion and close the PR:

1. Non-custom events on the loading screen are intentionally dropped to prevent a character from executing "old" events right after a warp.

2. Even if a non-custom event is dropped just before the loading screen appears, it's not a big deal, because a human player won't notice whether it was dropped right before or during the loading screen.

3. All event handlers in the engine should be able to handle dropped or partially dropped events (e.g. a key press consists of a "down" and an "up" events, and only the "up" event might get dropped because of point 1, thus "partially dropped" naming). Any bugs (like the one where the character keeps moving after a warp) should be fixed in their own handlers and are probably not caused by point 1 (at least there are no known such issues by now).

4. Any automation (like Lua scripts, etc.) has its own logic and needs to deal with dropped events (as described in point 1) on its own.

Points 1 and 4 are crucial for my automated agent, which makes me stick to my own branch. That's totally fine with me, I just wanted to clarify things.

Thanks to all of you folks for such a long conversation.